pkg/image/importer/client_test.go

```diff
@@ -0,0 +1,296 @@
+	return r.manifest, r.getByTagErr
+}
+
+func TestSchema1ToImage(t *testing.T) {
+	m := &schema1.SignedManifest{}
+	if err := json.Unmarshal([]byte(etcdManifest), m); err != nil {
+		t.Fatal(err)
+	}
+	image, err := schema1ToImage(m, digest.Digest("sha256:test"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if image.DockerImageMetadata.ID != "sha256:test" {
+		t.Errorf("unexpected image: %#v", image.DockerImageMetadata.ID)
+	}
+}
+
+func TestDockerV1Fallback(t *testing.T) {
+	var uri *url.URL
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("X-Docker-Endpoints", uri.Host)
+
+		// get all tags
+		if strings.HasSuffix(r.URL.Path, "/tags") {
+			fmt.Fprintln(w, `{"tag1":"image1", "test":"image2"}`)
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+		if strings.HasSuffix(r.URL.Path, "/images") {
+			fmt.Fprintln(w, `{"tag1":"image1", "test":"image2"}`)
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+		if strings.HasSuffix(r.URL.Path, "/json") {
+			fmt.Fprintln(w, `{"ID":"image2"}`)
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+		t.Logf("tried to access %s", r.URL.Path)
+		w.WriteHeader(http.StatusNotFound)
+	}))
+
+	client := dockerregistry.NewClient(10*time.Second, false)
+	ctx := gocontext.WithValue(gocontext.Background(), ContextKeyV1RegistryClient, client)
+
+	uri, _ = url.Parse(server.URL)
+	isi := &api.ImageStreamImport{
+		Spec: api.ImageStreamImportSpec{
+			Repository: &api.RepositoryImportSpec{
+				From:         kapi.ObjectReference{Kind: "DockerImage", Name: uri.Host + "/test:test"},
+				ImportPolicy: api.TagImportPolicy{Insecure: true},
+			},
+		},
+	}
+
+	retriever := &mockRetriever{err: fmt.Errorf("does not support v2 API")}
+	im := NewImageStreamImporter(retriever, 5, nil)
+	if err := im.Import(ctx, isi); err != nil {
+		t.Fatal(err)
+	}
+	if images := isi.Status.Repository.Images; len(images) != 2 || images[0].Tag != "tag1" || images[1].Tag != "test" {
+		t.Errorf("unexpected images: %#v", images)
+	}
+}
+
+func TestPing(t *testing.T) {
+	retriever := NewContext(http.DefaultTransport, http.DefaultTransport).WithCredentials(NoCredentials).(*repositoryRetriever)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	uri, _ := url.Parse(server.URL)
+
+	_, err := retriever.ping(*uri, true, retriever.context.InsecureTransport)
+	if !strings.Contains(err.Error(), "does not support v2 API") {
+		t.Errorf("Expected ErrNotV2Registry, got %v", err)
+	}
+
+	uri.Scheme = "https"
+	_, err = retriever.ping(*uri, true, retriever.context.InsecureTransport)
+	if !strings.Contains(err.Error(), "does not support v2 API") {
+		t.Errorf("Expected ErrNotV2Registry, got %v", err)
+	}
+}
+
+func TestShouldRetry(t *testing.T) {
+	r := NewRetryRepository(nil, 1, 0).(*retryRepository)
+
+	// nil error doesn't consume retries
+	if r.shouldRetry(nil) {
+		t.Fatal(r)
```

Comment:
Because there's no reason to continue the test if the condition is violated

On Tue, Apr 19, 2016 at 4:07 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/image/importer/client_test.go
> <https://github.com/openshift/origin/pull/8558#discussion_r60301490>:
>
> > +		t.Errorf("Expected ErrNotV2Registry, got %v", err)
> > +	}
> > +
> > +	uri.Scheme = "https"
> > +	_, err = retriever.ping(*uri, true, retriever.context.InsecureTransport)
> > +	if !strings.Contains(err.Error(), "does not support v2 API") {
> > +		t.Errorf("Expected ErrNotV2Registry, got %v", err)
> > +	}
> > +}
> > +
> > +func TestShouldRetry(t *testing.T) {
> > +	r := NewRetryRepository(nil, 1, 0).(*retryRepository)
> > +
> > +	// nil error doesn't consume retries
> > +	if r.shouldRetry(nil) {
> > +		t.Fatal(r)
>
> It's still unclear to me, why you're doing Fatalf here instead of just
> Errorf?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8558/files/e1c50b07d5369b627aa475724c2689dbdeb61a98#r60301490>
>